motoko/http_counter/src/main.mo

```diff
@@ -0,0 +1,120 @@
+import Nat "mo:base/Nat";
+import Text "mo:base/Text";
+import Array "mo:base/Array";
+import Option "mo:base/Option";
+import Prim "mo:prim";
+
+
+actor HttpCounter {
+
+  type StreamingCallbackHttpResponse = {
+    body: Blob;
+    token: ?Token;
+  };
+
+  type Token = {
+    key: Text;
+    content_encoding: Text;
+    index: Nat;
+    sha256: ?Blob;
+  };
+
+  type CallbackStrategy = {
+    callback: shared query (Token) -> async StreamingCallbackHttpResponse;
+    token: Token;
+  };
+
+  type StreamingStrategy =  {
+    #Callback: CallbackStrategy;
+  };
+
+  type HeaderField = (Text, Text);
+
+  type HttpResponse = {
+    status_code: Nat16;
+    headers: [HeaderField];
+    body: Blob;
+    streaming_strategy: ?StreamingStrategy;
+    upgrade: ?Bool;
+  };
+
+  type HttpRequest = {
+    method: Text;
+    url: Text;
+    headers: [HeaderField];
+    body: Blob;
+  };
+
+  func isGzip(x : HeaderField) : Bool {
+    Text.map(x.0 , Prim.charToLower) == "accept-encoding" and Text.contains(Text.map(x.1 , Prim.charToLower), #text "gzip");
+  };
+
+  stable var counter = 0;
+
+  public query func http_request(req : HttpRequest) : async HttpResponse {
+    switch (req.method, not Option.isNull(Array.find(req.headers, isGzip))) {
+      case ("GET", false) {{
+        status_code = 200;
+        headers = [ ("content-type", "text/plain") ];
+        body = Text.encodeUtf8("Counter is " # Nat.toText(counter) # "\n");
+        streaming_strategy = null;
+        upgrade = null;
+      }};
+      case ("GET", true) {{
+        status_code = 200;
+        headers = [ ("content-type", "text/plain"), ("content-encoding", "gzip") ];
+        body = "\1f\8b\08\00\98\02\1b\62\00\03\2b\2c\4d\2d\aa\e4\02\00\d6\80\2b\05\06\00\00\00";
```

Comment:
Does the gzipped variant not report the counter value? I guess you'd need a Motoko gzip library for that ...

Reply:
Ah, you say it's static in the README... nvm.